pkg/logs/decoder/line_handler.go

```diff
@@ -59,83 +66,88 @@ func (h *SingleLineHandler) Start() {
 	go h.run()
 }
 
-// run consumes lines from lineChan to process them
+// run consumes new lines and processes them.
 func (h *SingleLineHandler) run() {
 	for line := range h.lineChan {
 		h.process(line)
 	}
 	close(h.outputChan)
 }
 
-// process creates outputs from lines and forwards them to outputChan
-// When lines are too long, they are truncated
+// process transforms a raw line into a structured line,
+// it guarantees that the content of the line won't exceed
+// the limit and that the length of the line is properly tracked
+// so that the agent restarts tailing from the right place.
 func (h *SingleLineHandler) process(line []byte) {
-	lineLen := len(line)
-	line = bytes.TrimSpace(line)
-	if len(line) == 0 {
+	content, status, timestamp, err := h.parser.Parse(line)
+	if err != nil {
+		log.Debug(err)
+	}
+
+	isTruncated := h.shouldTruncate
+	h.shouldTruncate = false
+
+	rawLen := len(line)
+	if rawLen < h.lineLimit {
+		// lines are delimited on '\n' character when bellow the limit
+		// so we need to make sure it's properly accounted
+		rawLen++
+	}
+
+	content = bytes.TrimSpace(content)
```

Comment:
I would reorganize the code a bit to reduce interleaving of concerns:
```go
func (h *SingleLineHandler) process(line []byte) {
	isTruncated := h.shouldTruncate
	h.shouldTruncate = false

	rawLen := len(line)
	if rawLen < h.lineLimit {
		// lines are delimited on '\n' character when bellow the limit
		// so we need to make sure it's properly accounted
		rawLen++
	}

	content, status, timestamp, err := h.parser.Parse(line)
	if err != nil {
		log.Debug(err)
	}
	content = bytes.TrimSpace(content)
	if len(content) == 0 {
		// don't send empty lines
		return
	}
  (...)
}
```

Reply:
👍